libs/wingc/src/inflight_transform.rs

```diff
@@ -0,0 +1,277 @@
+/// For example, the following code:
+/// ```wing
+/// let b = new cloud.Bucket();
+/// let f = inflight (message: str) => {
+///   b.put("file.txt", message);
+/// };
+/// ```
+/// is transformed into:
+/// ```wing
+/// let b = new cloud.Bucket();
+/// let f = ((): core.Resource => {
+///   class $Inflight1 {
+///     init() {}
+///     inflight handle(message: str) {
+///       b.put("file.txt", message);
+///     }
+///   }
+///   return new $Inflight1();
+/// })();
+pub struct InflightTransformer {
+	// Whether the transformer is inside a preflight or inflight scope.
+	// Only inflight closures defined in preflight scopes need to be transformed.
+	curr_phase: Phase,
+	// Whether the transformer is inside a resource or not. If we are, we transform
+	// inflight closures with an extra `let __parent_this = this;` statement so
+	// that they can access the parent resource's fields.
+	is_inside_resource: bool,
+	// Helper state for generating unique class names for resources
+	inflight_counter: usize,
+}
+
+impl InflightTransformer {
+	pub fn new() -> Self {
+		Self {
+			curr_phase: Phase::Preflight,
+			is_inside_resource: false,
+			inflight_counter: 0,
+		}
+	}
+}
+
+impl Fold for InflightTransformer {
+	fn fold_function_definition(&mut self, node: FunctionDefinition) -> FunctionDefinition {
+		let prev_phase = self.curr_phase;
+		self.curr_phase = node.signature.phase;
+		let new_node = fold::fold_function_definition(self, node);
+		self.curr_phase = prev_phase;
+		new_node
+	}
+
+	fn fold_class(&mut self, node: Class) -> Class {
+		let prev_is_inside_resource = self.is_inside_resource;
+		self.is_inside_resource = true;
+		let new_node = fold::fold_class(self, node);
+		self.is_inside_resource = prev_is_inside_resource;
+		new_node
+	}
+
+	fn fold_expr(&mut self, expr: Expr) -> Expr {
+		// Inflight closures that are themselves defined inflight do not need
+		// to be transformed, since they are not created in preflight.
+		if self.curr_phase == Phase::Inflight {
+			return expr;
+		}
+
+		// If we encounter a non-inflight closure, we can don't need to
+		// transform it into a resource, but we still want to recurse
+		// in case its body contains any inflight closures.
+		if let ExprKind::FunctionClosure(ref func_def) = expr.kind {
+			if func_def.signature.phase != Phase::Inflight {
+				return fold::fold_expr(self, expr);
+			}
+		}
+
+		match expr.kind {
+			ExprKind::FunctionClosure(func_def) => {
+				self.inflight_counter += 1;
+
+				let resource_name = Symbol {
+					name: format!("$Inflight{}", self.inflight_counter),
+					span: expr.span.clone(),
+				};
+				let handle_name = Symbol {
+					name: "handle".to_string(),
+					span: expr.span.clone(),
+				};
+				let parent_this_name = Symbol {
+					name: PARENT_THIS_NAME.to_string(),
+					span: expr.span.clone(),
+				};
+				let this_name = Symbol {
+					name: "this".to_string(),
+					span: expr.span.clone(),
+				};
+
+				let resource_type_annotation = TypeAnnotation {
```

Comment:
```suggestion
				let class_type = TypeAnnotation {
```


Why do we call this TypeAnnotation??

Reply:
This is to help us distinguish them from the `Type` structs we generate and store during type checking. For example, when an expression node is type checked, its `evaluated_type` is assigned a value with the type it has.

Part of the reason this is needed in the current compiler architecture is because we need a way to represent code like `let x: foo.Bar = // snip` before we have started resolving symbols or doing type checking (steps we currently do together).